File: app/src/main/java/com/kerimovscreations/lateandroid/dialogs/SettingsDialogFragment.java
package com.kerimovscreations.lateandroid.dialogs;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.Context;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import android.view.View;
import android.view.Window;
import android.widget.TextView;

import com.kerimovscreations.lateandroid.R;
import com.kerimovscreations.lateandroid.application.GlobalApplication;
import com.kerimovscreations.lateandroid.enums.SoundType;
import com.kerimovscreations.lateandroid.tools.HelpFunctions;
import com.kerimovscreations.lateandroid.tools.LocaleHelper;

import java.util.Objects;

public class SettingsDialogFragment extends Dialog implements
        android.view.View.OnClickListener {

    private Context mContext;
    private TextView mLanguageTitle;
    private TextView mLanguageText;
    private TextView mSoundTitle;
    private TextView mSoundText;
    private TextView mTitle;
    private TextView mSubmitBtn;

    public SettingsDialogFragment(Context context) {
        super(context);
        mContext = context;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        requestWindowFeature(Window.FEATURE_NO_TITLE);
        setContentView(R.layout.dialog_settings);

        Objects.requireNonNull(getWindow()).setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));

        findViewById(R.id.btn_submit).setOnClickListener(this);
        findViewById(R.id.language_view).setOnClickListener(this);
        findViewById(R.id.sound_view).setOnClickListener(this);

        mTitle = findViewById(R.id.title);
        mLanguageTitle = findViewById(R.id.language_title);
        mLanguageText = findViewById(R.id.language_text);
        mSoundTitle = findViewById(R.id.sound_title);
        mSoundText = findViewById(R.id.sound_text);
        mSubmitBtn = findViewById(R.id.btn_submit);
        updateTexts();
    }

    private void updateTexts() {
        mTitle.setText(mContext.getString(R.string.settings));
        mLanguageTitle.setText(mContext.getString(R.string.language));
        mLanguageText.setText(mContext.getString(R.string.current_language));
        mSoundTitle.setText(mContext.getString(R.string.sound));

        SoundType soundType = SoundType.values()[HelpFunctions.shared.getSoundType(mContext)];

        int resId;
        switch (soundType) {
            case MALE_NORMAL:
                resId = R.string.male;
                break;
            case FEMALE_NORMAL:
                resId = R.string.female;
                break;
            case MALE_FUNNY_1:
                resId = R.string.male_fun_1;
                break;
            case FEMALE_FUNNY_1:
                resId = R.string.female_fun_1;
                break;
            case MALE_FUNNY_2:
                resId = R.string.male_fun_2;
                break;
            case FEMALE_FUNNY_2:
                resId = R.string.female_fun_2;
                break;
            default:
                resId = R.string.male;
        }

        mSoundText.setText(mContext.getString(resId));

        mSubmitBtn.setText(mContext.getString(R.string.submit));
    }

    /* Click handlers */

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.btn_submit:
                dismiss();
                break;
            case R.id.language_view:
                onLanguage();
                break;
            case R.id.sound_view:
                onSound();
                break;
            default:
                break;
        }
    }

    private void onLanguage() {
        AlertDialog.Builder adb = new AlertDialog.Builder(mContext, R.style.AppThemeAlertDialog);
        CharSequence[] items = new CharSequence[]{"English", "Русский"};
        final int[] selectedOption = {-1};

        if (GlobalApplication.localeManager.getLanguage().equals("en")) {
            selectedOption[0] = 0;
        } else if (GlobalApplication.localeManager.getLanguage().equals("ru")) {
            selectedOption[0] = 1;
        } else {
            selectedOption[0] = 0;
        }

        adb.setSingleChoiceItems(items, selectedOption[0], (dialog, which) -> selectedOption[0] = which);
        adb.setPositiveButton(mContext.getString(R.string.save), (dialog, which) -> {
            switch (selectedOption[0]) {
                case 0:
                    GlobalApplication.localeManager.setNewLocale(mContext, "en");
                    HelpFunctions.shared.setUserLanguage(mContext, "en");
                    updateTexts();
                    break;
                case 1:
                    GlobalApplication.localeManager.setNewLocale(mContext, "ru");
                    HelpFunctions.shared.setUserLanguage(mContext, "ru");
                    updateTexts();
                    break;
                default:
                    break;
            }
        });
        adb.setNegativeButton(mContext.getString(R.string.cancel), null);
        adb.setTitle(mContext.getString(R.string.select_language));
        adb.show();
    }

    private void onSound() {
        AlertDialog.Builder adb = new AlertDialog.Builder(mContext, R.style.AppThemeAlertDialog);
        CharSequence[] items;

        if (GlobalApplication.localeManager.getLanguage().equals("en")) {
            items = new CharSequence[]{mContext.getString(R.string.male), mContext.getString(R.string.female)};
        } else {
            items = new CharSequence[]{mContext.getString(R.string.male), mContext.getString(R.string.female),
                    mContext.getString(R.string.male_fun_1), mContext.getString(R.string.female_fun_1),
                    mContext.getString(R.string.male_fun_2), mContext.getString(R.string.female_fun_2)};
        }

        final int[] selectedOption = {-1};

        SoundType soundType = SoundType.values()[HelpFunctions.shared.getSoundType(mContext)];
        selectedOption[0] = soundType.getValue();

        adb.setSingleChoiceItems(items, selectedOption[0], (dialog, which) -> selectedOption[0] = which);
        adb.setPositiveButton(mContext.getString(R.string.save), (dialog, which) -> {
            HelpFunctions.shared.setSoundType(mContext, selectedOption[0]);
            updateTexts();
        });
        adb.setNegativeButton(mContext.getString(R.string.cancel), null);
        adb.setTitle(mContext.getString(R.string.select_sound));
        adb.show();
    }
}

File: app/src/main/java/com/kerimovscreations/lateandroid/dialogs/SettingsDialogFragment.kt
package com.kerimovscreations.lateandroid.dialogs

import android.app.AlertDialog
import android.content.DialogInterface
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.DialogFragment
import com.kerimovscreations.lateandroid.R
import com.kerimovscreations.lateandroid.application.GlobalApplication
import com.kerimovscreations.lateandroid.databinding.DialogSettingsBinding
import com.kerimovscreations.lateandroid.enums.SoundType
import com.kerimovscreations.lateandroid.tools.HelpFunctions

class SettingsDialogFragment : DialogFragment() {

    companion object {
        fun newInstance() = SettingsDialogFragment()
    }

    private lateinit var binding: DialogSettingsBinding

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding =
            DataBindingUtil.inflate(inflater, R.layout.dialog_settings, container, false)

        binding.btnSubmit.setOnClickListener {
            dismiss()
        }

        binding.languageView.setOnClickListener {
            onLanguage()
        }

        binding.soundView.setOnClickListener {
            onSound()
        }

        updateTexts()

        return binding.root
    }

    private fun updateTexts() {
        binding.title.text = getString(R.string.settings)
        binding.languageTitle.text = getString(R.string.language)
        binding.languageText.text = when (GlobalApplication.localeManager!!.language) {
            "en" -> "English"
            "ru" -> "Русский"
            else -> "English"
        }
        binding.soundTitle.text = getString(R.string.sound)
        val soundType =
            SoundType.entries.getOrElse(
                HelpFunctions.shared.getSoundType(this.requireContext()),
                { SoundType.MALE_NORMAL })
        val resId: Int = when (soundType) {
            SoundType.MALE_NORMAL -> R.string.male
            SoundType.FEMALE_NORMAL -> R.string.female
            SoundType.MALE_FUNNY_1 -> R.string.male_fun_1
            SoundType.FEMALE_FUNNY_1 -> R.string.female_fun_1
            SoundType.MALE_FUNNY_2 -> R.string.male_fun_2
            SoundType.FEMALE_FUNNY_2 -> R.string.female_fun_2
        }

        binding.soundText.text = getString(resId)
        binding.btnSubmit.text = getString(R.string.submit)
    }

    private fun onLanguage() {
        val adb = AlertDialog.Builder(this.requireContext(), R.style.AppThemeAlertDialog)
        val items = arrayOf<CharSequence>("English", "Русский")
        var selectedOption = when (GlobalApplication.localeManager!!.language) {
            "en" -> {
                0
            }

            "ru" -> {
                1
            }

            else -> {
                0
            }
        }
        adb.setSingleChoiceItems(
            items,
            selectedOption
        ) { _: DialogInterface?, which: Int -> selectedOption = which }
        adb.setPositiveButton(getString(R.string.save)) { _: DialogInterface?, _: Int ->
            when (selectedOption) {
                0 -> {
                    GlobalApplication.localeManager!!.setNewLocale(this.requireContext(), "en")
                    HelpFunctions.shared.setUserLanguage(this.requireContext(), "en")
                    HelpFunctions.shared.setSoundType(this.requireContext(), SoundType.MALE_NORMAL)
                    updateTexts()
                }

                1 -> {
                    GlobalApplication.localeManager!!.setNewLocale(this.requireContext(), "ru")
                    HelpFunctions.shared.setUserLanguage(this.requireContext(), "ru")
                    HelpFunctions.shared.setSoundType(this.requireContext(), SoundType.MALE_NORMAL)
                    updateTexts()
                }

                else -> {
                }
            }
        }
        adb.setNegativeButton(getString(R.string.cancel), null)
        adb.setTitle(getString(R.string.select_language))
        adb.show()
    }

    private fun onSound() {
        val adb = AlertDialog.Builder(this.requireContext(), R.style.AppThemeAlertDialog)
        val language = GlobalApplication.localeManager!!.language
        val items: Array<Pair<SoundType, CharSequence>> = if (language == "en") {
            arrayOf(
                Pair(SoundType.MALE_NORMAL, getString(R.string.male)),
                Pair(SoundType.FEMALE_NORMAL, getString(R.string.female)),
            )
        } else {
            arrayOf(
                Pair(SoundType.MALE_NORMAL, getString(R.string.male)),
                Pair(SoundType.FEMALE_NORMAL, getString(R.string.female)),
                Pair(SoundType.MALE_FUNNY_1, getString(R.string.male_fun_1)),
                Pair(SoundType.FEMALE_FUNNY_1, getString(R.string.female_fun_1)),
                Pair(SoundType.MALE_FUNNY_2, getString(R.string.male_fun_2)),
                Pair(SoundType.FEMALE_FUNNY_2, getString(R.string.female_fun_2)),
            )
        }
        var selectedOption: Int
        val soundType = SoundType.entries.getOrElse(
            HelpFunctions.shared.getSoundType(this.requireContext()),
            { SoundType.MALE_NORMAL })
        selectedOption = items.firstOrNull { it.first == soundType }?.let { items.indexOf(it) } ?: 0
        adb.setSingleChoiceItems(
            items.map { it.second }.toTypedArray(),
            selectedOption
        ) { _: DialogInterface?, which: Int -> selectedOption = which }
        adb.setPositiveButton(getString(R.string.save)) { _: DialogInterface?, _: Int ->
            HelpFunctions.shared.setSoundType(
                this.requireContext(),
                items[selectedOption].first
            )
            updateTexts()
        }
        adb.setNegativeButton(getString(R.string.cancel), null)
        adb.setTitle(getString(R.string.select_sound))
        adb.show()
    }
}